Implement BIO.parMap2 ... parMap6 and parZip

Executioner1939 opened the pull request:
#26 

@Avasil this one is ready for review :-).

Executioner1939 commented:
@Avasil Hi 😄 

Anything I can do still? Please let me know 😄 😄

Avasil commented on core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:
```diff
@@ -0,0 +1,291 @@
+
+  test("BIO.parZip2 should onError from the source before other and return the user error") { implicit s =>
+    val ex0 = DummyException("dummy")
+    val ex1 = DummyErrorResponse("dummy")
```
I feel like `DummyErrorResponse` class is unnecessary, we can use `DummyException` with different message instead

Executioner1939 replied:
Agreed, I will remove this :-).

Avasil commented on core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:
```diff
@@ -0,0 +1,291 @@
+    val ex1 = DummyErrorResponse("dummy")
+
+    val f = BIO
+      .parZip2(BIO(throw ex0).mapError(_ => ex1).delayExecution(2.second), BIO(2).delayExecution(1.seconds))
```
IMO `mapError`s in these tests are not relevant to `parZip` and could be removed

Executioner1939 replied:
Hi Avasil :-). I see I completely misunderstood, I thought thow would result in a terminal error. This test is redundant then.

I have gone through the runloop and seen that the Eval and Suspend Nodes catch the Error and raise it to the non-terminal error channel.

Avasil replied:
`BIO.apply` which is alias to `eval` catches any errors and that's why it returns `BIO[Throwable, A]`

It doesn't happen for other methods, e.g. `flatMap(_ => throw ex)` won't be caught in typed error channel.

There is also `evalTotal` (and UIO.apply) which returns `UIO` so any error thrown there would be terminal

Executioner1939 commented:
@Avasil I updated the PR, removed the `DummyErrorResponse`, removed the tests using `mapError` and added two using `BIO.terminate`, one for each "side" of `parZip2`.
